MREUnityRuntime/MREUnityRuntimeLib/Assets/AssetManager.cs

```diff
@@ -0,0 +1,318 @@
+		/// a copy of the asset will be made, and stored back into the manager. Any other shared assets that reference
+		/// this asset will also be recursively copied and stored back. Each copied asset will have the original
+		/// returned to the cache, decrementing the original's reference count.</param>
+		/// <returns></returns>
+		public AssetMetadata? GetById(Guid? id, bool writeSafe = false)
+		{
+			if (id != null && Assets.TryGetValue(id.Value, out AssetMetadata metadata))
+			{
+				// copy sourced assets if requesting write-safe
+				if (writeSafe)
+				{
+					MakeWriteSafe(metadata);
+				}
+				return Assets[id.Value];
+			}
+			else return null;
+		}
+
+		/// <summary>
+		/// Retrieve an asset's metadata from the asset reference itself.
+		/// </summary>
+		/// <param name="asset"></param>
+		/// <returns></returns>
+		public AssetMetadata? GetByObject(Object asset)
+		{
+			foreach (var metadata in Assets.Values)
+			{
+				if (metadata.Asset == asset)
+				{
+					return metadata;
+				}
+			}
+			return null;
+		}
+
+		/// <summary>
+		/// Be notified when an asset is finished loading and available for use.
+		/// </summary>
+		/// <param name="id"></param>
+		/// <param name="callback"></param>
+		public void OnSet(Guid id, AssetCallback callback)
+		{
+			var asset = GetById(id);
+			if (asset != null)
+			{
+				try
+				{
+					callback?.Invoke(asset.Value);
+				}
+				catch (Exception e)
+				{
+					Debug.LogException(e);
+				}
+			}
+			else
+			{
+				Callbacks.GetOrCreate(id, () => new List<AssetCallback>(10)).Add(callback);
+			}
+		}
+
+		/// <summary>
+		/// Track a new asset reference. Will be called during asset creation, after the asset content is downloaded
+		/// or retrieved from cache.
+		/// </summary>
+		/// <param name="id"></param>
+		/// <param name="containerId"></param>
+		/// <param name="asset"></param>
+		/// <param name="colliderGeo"></param>
+		/// <param name="source"></param>
+		public void Set(Guid id, Guid containerId, Object asset,
+			ColliderGeometry colliderGeo = null, AssetSource source = null)
+		{
+			if (!Assets.ContainsKey(id))
+			{
+				Assets[id] = new AssetMetadata(id, containerId, asset, colliderGeo, source);
+			}
+
+			if (Callbacks.TryGetValue(id, out List<AssetCallback> callbacks))
+			{
+				Callbacks.Remove(id);
+				foreach (var cb in callbacks)
+				{
+					try
+					{
+						cb?.Invoke(Assets[id]);
+					}
+					catch(Exception e)
+					{
+						Debug.LogException(e);
+					}
+				}
+			}
+		}
+
+		/// <summary>
+		/// Break references to all shared assets and destroy all unshared assets with this container ID.
+		/// </summary>
+		/// <param name="containerId"></param>
+		public void Unload(Guid containerId)
+		{
+			var assets = Assets.Values.Where(c => c.ContainerId == containerId && c.Asset != null).ToArray();
+			for (var i = 0; i < assets.Length; i++)
+			{
+				var asset = assets[i];
```

Comment:
This can be simplified to a ranged foreach loop since i  doesn't appear to be needed for anything other than indexing the current asset for the loop.